libvast/vast/flatbuffer.hpp

```diff
@@ -0,0 +1,350 @@
+
+  /// Constructs a ref-counted FlatBuffers root table that shares the
+  /// lifetime with the chunk it's constructed from.
+  /// @pre *chunk* must hold a valid *Table*.
+  [[nodiscard]] static caf::expected<flatbuffer>
+  make(const chunk_ptr& chunk, enum verify verify = verify_default) noexcept
+    requires(IsRootTable) {
+    if (!chunk)
+      return caf::make_error(ec::logic_error,
+                             fmt::format("failed to read {} from a nullptr",
+                                         Table::GetFullyQualifiedName()));
+    if (chunk->size() == 0)
+      return caf::make_error(
+        ec::logic_error, fmt::format("failed to read {} from an empty chunk",
+                                     Table::GetFullyQualifiedName()));
+    if (chunk->size() >= FLATBUFFERS_MAX_BUFFER_SIZE)
+      return caf::make_error(
+        ec::format_error,
+        fmt::format("failed to read {} because its size {} "
+                    "exceeds the maximum allowed size of {}",
+                    Table::GetFullyQualifiedName(), chunk->size(),
+                    FLATBUFFERS_MAX_BUFFER_SIZE));
+    if (verify == verify::yes) {
+      const auto* const data = reinterpret_cast<const uint8_t*>(chunk->data());
+      auto verifier = flatbuffers::Verifier{data, chunk->size()};
+      if (!flatbuffers::GetRoot<Table>(data)->Verify(verifier))
+        return caf::make_error(ec::format_error,
+                               fmt::format("failed to read {} because its "
+                                           "verification failed",
+                                           Table::GetFullyQualifiedName()));
+#if defined(FLATBUFFERS_TRACK_VERIFIER_BUFFER_SIZE)
+      VAST_ASSERT(verifier.GetComputedSize() >= chunk->size());
+      if (verifier.GetComputedSize() > chunk->size())
+        return caf::make_error(
+          ec::format_error,
+          fmt::format("failed to read {} because of {} unexpected excess bytes",
+                      Table::GetFullyQualifiedName(),
+                      verifier.GetComputedSize() - chunk->size()));
+#endif // defined(FLATBUFFERS_TRACK_VERIFIER_BUFFER_SIZE)
+    }
+    return flatbuffer{chunk};
+  }
+
+  /// Constructs a ref-counted FlatBuffers root table.
+  /// @pre *buffer* must hold a valid *Table*.
+  [[nodiscard]] static caf::expected<flatbuffer>
+  make(flatbuffers::DetachedBuffer&& buffer, enum verify verify
+                                             = verify_default) noexcept {
+    return make(chunk::make(std::move(buffer)), verify);
+  }
+
+  /// Default-constructs a FlatBuffers table.
+  flatbuffer() noexcept = default;
+
+  /// Constructs a ref-counted FlatBuffers root table from a FlatBufferBuilder
+  /// by finishing it.
+  // TODO: FlatBuffers 2.0's reflection allows for inferring the
+  // *file_identifier* from *Table*, which makes this API a lot easier to use.
+  // When upgrading to that version from 1.12 we can enable reflection when
+  // compiling the schemas.
```

Comment:
In a call we discussed instead introducing a `flatbuffer_traits<T>::file_identifier` map for all our root types, since the correct file identifier is already determined by `Table`, even though flatbuffers 1.12 lacks the API to figure it out.

Reply:
I've tried this, and it's held back by the fact that this needs the definition of `T` available, which we don't want to include in this header. So I think it's best to leave this as-is for now.